### PR TITLE
 Add (low level) interface to `lrs_solve_lp`

### PR DESCRIPTION
**Pull request by oyamad**
To solve

min -x s.t. 7x + y <= 3/2, x, y >= 0:

```jl
using Polyhedra
using LRSLib: RepMatrix, setobj, lpsolve

a = [7, 1]
b = 3//2
c = [-1//1, 0]
hr = hrep([HalfSpace(a, b), HalfSpace([-1//1, 0], 0), HalfSpace([0, -1//1], 0)])
maximize = false

m = RepMatrix(hr)
setobj(m, c, maximize)
sol = lpsolve(m)

println("Optimal objective value is: $(sol.objval)")
println("Optimal solution vector is: [$(sol.sol[1]), $(sol.sol[2])]")
```

```
Optimal objective value is: -3//14
Optimal solution vector is: [3//14, 0//1]
```

I am not knowledgable enough to implement interface through `MathProgBase`. I hope someone (most probably @blegat) could do in the future.

### CODE REVIEW

File: src/LRSLib.jl [hunk 1]
```diff
@@ -2,6 +2,7 @@ module LRSLib
 
 using BinDeps
 using Polyhedra
+using Markdown
```
Comment (blegat) [inline src/LRSLib.jl]: Why is this needed ? What does it do ?

File: src/conversion.jl [hunk 2]
```diff
@@ -26,7 +26,7 @@ function generatorproducer(m::RepMatrix)
             while true
                 for col in 0:getd(m)
                     output = getsolution(m, col)
-                    if output !== nothing
+                    if length(output) != 0
```
Comment (blegat) [inline src/conversion.jl]: Was this a bug ?

File: src/lp.jl [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+setobj(m::HMatrix, c::Vector{<:Real}, maximize::Bool) =
+    setobj(m.P, m.Q, c, maximize)
+
+@doc """
+    setobj(m::HMatrix, c::Vector{<:Real}, maximize::Boo)
+
+Set the objective function vector `c` for `m`. Set `maximize` to true (false,
+resp.) if the objective is to be maximized (minimized, resp.).
+""" setobj
```
Comment (blegat) [inline src/lp.jl]: Doesn't it need to be right before `set_obj` ?

File: src/matrix.jl [hunk 4]
```diff
@@ -392,7 +392,7 @@ function getsolution(m::RepMatrix, col::Int)
     if found
         out = convertoutput(output, Q.n, Q.hull == Clrs_true)
     else
-        out = nothing
+        out = Rational{BigInt}[]
```
Comment (blegat) [inline src/matrix.jl]: Why is this needed ?

File: src/lp.jl (hunk 3 above)
Comment (oyamad) [reply src/lp.jl]: It seems to work anywhere with `@doc`. Refer to [this](https://docs.julialang.org/en/v1/manual/documentation/#Advanced-Usage-1).

File: src/LRSLib.jl (hunk 1 above)
Comment (oyamad) [reply src/LRSLib.jl]: This is needed for the docstring with `@doc doc` for `lpsolve` (which uses LaTeX commands).

File: src/matrix.jl (hunk 4 above)
Comment (oyamad) [reply src/matrix.jl]: This is not really needed. It is to resolve the type instability of `getsolution`.

File: src/conversion.jl (hunk 2 above)
Comment (oyamad) [reply src/conversion.jl]: No. Please see my reply to your comment on `getsolution`.

File: src/matrix.jl (hunk 4 above)
Comment (oyamad) [reply src/matrix.jl]: I didn't know that. I just felt scared by the red color in the output of `@code_warntype`. I can revert the second commit if its change is inappropriate.

Comment (blegat) [reply src/matrix.jl]: I think it is better to revert it. See the following benchmark:
```
julia> f() = Rational{BigInt}[]
f (generic function with 1 metjulia> f() = Rational{BigInt}[]
f (generic function with 1 method)

julia> g() = nothing
g (generic function with 1 method)

julia> t(f) = @time f()
t (generic function with 1 method)

julia> h() = isempty(f())
h (generic function with 1 method)

julia> i() = g() === nothing
i (generic function with 1 method)

julia> t(h)
  0.000001 seconds (1 allocation: 80 bytes)
true

julia> t(h)
  0.000000 seconds (1 allocation: 80 bytes)
true

julia> t(i)
  0.000000 seconds
true

julia> t(i)
  0.000000 seconds
true

julia> using BenchmarkTools
[ Info: Recompiling stale cache file /home/blegat/.julia/compiled/v1.0/BenchmarkTools/ZXPQo.ji for BenchmarkTools [6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf]

julia> @btime h()
  19.100 ns (1 allocation: 80 bytes)
true

julia> @btime i()
  0.025 ns (0 allocations: 0 bytes)
true

julia> @btime i()
  0.026 ns (0 allocations: 0 bytes)
truejulia> f() = Rational{BigInt}[]
f (generic function with 1 method)

julia> g() = nothing
g (generic function with 1 method)

julia> t(f) = @time f()
t (generic function with 1 method)

julia> h() = isempty(f())
h (generic function with 1 method)

julia> i() = g() === nothing
i (generic function with 1 method)

julia> t(h)
  0.000001 seconds (1 allocation: 80 bytes)
true

julia> t(h)
  0.000000 seconds (1 allocation: 80 bytes)
true

julia> t(i)
  0.000000 seconds
true

julia> t(i)
  0.000000 seconds
true

julia> using BenchmarkTools
[ Info: Recompiling stale cache file /home/blegat/.julia/compiled/v1.0/BenchmarkTools/ZXPQo.ji for BenchmarkTools [6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf]

julia> @btime h()
  19.100 ns (1 allocation: 80 bytes)
true

julia> @btime i()
  0.025 ns (0 allocations: 0 bytes)
true

julia> @btime i()
  0.026 ns (0 allocations: 0 bytes)
truejulia> f() = Rational{BigInt}[]
f (generic function with 1 method)

julia> g() = nothing
g (generic function with 1 method)

julia> t(f) = @time f()
t (generic function with 1 method)

julia> h() = isempty(f())
h (generic function with 1 method)

julia> i() = g() === nothing
i (generic function with 1 method)

julia> t(h)
  0.000001 seconds (1 allocation: 80 bytes)
true

julia> t(h)
  0.000000 seconds (1 allocation: 80 bytes)
true

julia> t(i)
  0.000000 seconds
true

julia> t(i)
  0.000000 seconds
true

julia> using BenchmarkTools
[ Info: Recompiling stale cache file /home/blegat/.julia/compiled/v1.0/BenchmarkTools/ZXPQo.ji for BenchmarkTools [6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf]

julia> @btime h()
  19.100 ns (1 allocation: 80 bytes)
true

julia> @btime i()
  0.025 ns (0 allocations: 0 bytes)
true

julia> @btime i()
  0.026 ns (0 allocations: 0 bytes)
truehod)
julia> f() = Rational{BigInt}[]
f (generic function with 1 method)

julia> g() = nothing
g (generic function with 1 method)

julia> t(f) = @time f()
t (generic function with 1 method)

julia> h() = isempty(f())
h (generic function with 1 method)

julia> i() = g() === nothing
i (generic function with 1 method)

julia> t(h)
  0.000001 seconds (1 allocation: 80 bytes)
true

julia> t(h)
  0.000000 seconds (1 allocation: 80 bytes)
true

julia> t(i)
  0.000000 seconds
true

julia> t(i)
  0.000000 seconds
true

julia> using BenchmarkTools
[ Info: Recompiling stale cache file /home/blegat/.julia/compiled/v1.0/BenchmarkTools/ZXPQo.ji for BenchmarkTools [6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf]

julia> @btime h()
  19.100 ns (1 allocation: 80 bytes)
true

julia> @btime i()
  0.025 ns (0 allocations: 0 bytes)
true

julia> @btime i()
  0.026 ns (0 allocations: 0 bytes)
true
julia> g() = nothing
g (generic function with 1 method)

julia> t(f) = @time f()
t (generic function with 1 method)

julia> h() = isempty(f())
h (generic function with 1 method)

julia> i() = g() === nothing
i (generic function with 1 method)

julia> t(h)
  0.000001 seconds (1 allocation: 80 bytes)
true

julia> t(h)
  0.000000 seconds (1 allocation: 80 bytes)
true

julia> t(i)
  0.000000 seconds
true

julia> t(i)
  0.000000 seconds
true

julia> using BenchmarkTools
[ Info: Recompiling stale cache file /home/blegat/.julia/compiled/v1.0/BenchmarkTools/ZXPQo.ji for BenchmarkTools [6e4b80f9-dd63-53aa-95a3-0cdb28fa8baf]

julia> @btime h()
  19.100 ns (1 allocation: 80 bytes)
true

julia> @btime i()
  0.025 ns (0 allocations: 0 bytes)
true

julia> @btime i()
  0.026 ns (0 allocations: 0 bytes)
true
```

Comment (oyamad) [reply src/matrix.jl]: Confirmed there is no significant difference:

```jl
f(flag) = flag ? [BigInt(1)//1] : Rational{BigInt}[]  # type stable
g(flag) = flag ? [BigInt(1)//1] : nothing  # type unsable

function F(flags)
    out = Vector{Rational{BigInt}}[]
    for i in eachindex(flags)
        a = f(flags[i])
        if !isempty(a)
            push!(out, a)
        end
    end
    out
end

function G(flags)
    out = Vector{Rational{BigInt}}[]
    for i in eachindex(flags)
        a = g(flags[i])
        if a !== nothing
            push!(out, a)
        end
    end
    out
end

flags = rand(Bool, 10^4);
```

```jl
@btime F(flags);
```

```
  3.982 ms (116525 allocations: 3.06 MiB)
```

```jl
@btime G(flags);
```

```
  3.804 ms (111597 allocations: 2.68 MiB)
```

**Comment by oyamad** [comment]
Deleted the second commit (and made a force push).

File: src/lp.jl [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+mutable struct LRSLinprogSolution
+    status::Symbol  # :Optimal, :Infeasible, :Unbounded, :UserLimit, or :Error
+    objval::Union{Nothing,Rational{BigInt}}
+    sol::Vector{Rational{BigInt}}
+    attrs::Dict
+end
+
+
+function setobj(P::Ptr{Clrs_dic}, Q::Ptr{Clrs_dat},
+                c::Vector{<:Real}, maximize::Bool)
```
Comment (blegat) [inline src/lp.jl]: What do you think of changing `<:Real` to `<:Union{Rational, Integer}` ? If `c` is a vector of `Float64`, the user will get a bad surprise and will prefer a method error IMO.

File: src/lp.jl [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+    row = Vector{Rational{BigInt}}(undef, unsafe_load(P).d+1)
+    sgn = maximize ? 1 : -1
+    field = maximize ? :maximize : :minimize
+    row[1] = 0
```
Comment (blegat) [inline src/lp.jl]: `row[1] = zero(eltype(row))` might be better as `convert(Rational{BigInt}, 0)` will call a constructor that is not specific to zero and might be less appropriate

File: src/lp.jl [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    row[2:end] = c * sgn
```
Comment (blegat) [inline src/lp.jl]: I would do `= maximize ? sgn : -sgn`. If `maximize` is `true` it will not do any multiplication which is more efficient

File: src/lp.jl [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    for col in 0:(Q.n)-1
+        found = (
+            Clrs_true ==
+            @lrs_ccall getsolution Clong (Ptr{Clrs_dic}, Ptr{Clrs_dat}, Clrs_mp_vector, Clong) m.P m.Q output col
+        )
+        if print_solution
+            if found
+                println("col=$(col): found")
+                @lrs_ccall printoutput Cvoid (Ptr{Clrs_dat}, Clrs_mp_vector) m.Q output
+                print("\n")
```
Comment (blegat) [inline src/lp.jl]: Why not `println()` ?

File: src/lp.jl [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    println("*Objective function has value $(num)/$(den)")
+
+    print("*Primal:")
+    for i in 1:Q.n-1
+        print(" x_$i=")
+        num = extractbigint(unsafe_load(output, i+1))
+        den = extractbigint(unsafe_load(output, 1))
+        print(num, "/", den)
+    end
+    print("\n")
```
Comment (blegat) [inline src/lp.jl]: Again, why not `println` ?

File: src/lp.jl [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        Col_i = unsafe_load(P.Col, i+1)
+        idx = unsafe_load(Q.inequality, C_i-Q.lastdv+1)
+        print(" y_$(idx)=")
+        temp1 = extractbigint(unsafe_load(Q.Lcm, Col_i+1)) * (-1)
+        temp1 *= extractbigint(unsafe_load(unsafe_load(P.A, 1), Col_i+1))
+        temp2 = extractbigint(unsafe_load(Q.Gcd, Col_i+1))
+        temp2 *= extractbigint(P.det)
+        print(temp1, "/", temp2)
+    end
+    print("\n")
```
Comment (blegat) [inline src/lp.jl]: ditto

**Comment by oyamad** [comment]
@blegat Thanks for your comments. Incorporated them all.
